### PR TITLE
[dockerfile] remove build deps from final image

### PR DESCRIPTION
**Pull request by alebabai**
## what
remove build dependencies from final image when all `make install` steps are done

## why
#30

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -3,7 +3,8 @@
 ##
 FROM alpine:3.8 as builder
 
-RUN apk --update add --virtual .build-deps build-base automake autoconf libtool git linux-pam-dev openssl-dev wget
+ENV BUILD_DEPENDENCIES=".build-deps build-base automake autoconf libtool git linux-pam-dev openssl-dev wget"
```
Comment (osterman) [inline Dockerfile]: You don't need this env

File: Dockerfile [hunk 2]
```diff
@@ -94,6 +95,9 @@ COPY --from=openssh-portable-builder dist dist
 RUN make --directory=dist install && \
     rm -rf dist
 
+## Remove build dependencies
+RUN apk del ${BUILD_DEPENDENCIES}
```
Comment (osterman) [inline Dockerfile]: Just delete the `.build-deps` virtual package

**Comment by osterman** [comment]
My concern with this is we're not really optimizing the image size by doing the multi-stage. Can you try just adding `make` as a dep from a clean alpine base and seeing if you can successfully install?

**Comment by osterman** [comment]
> in this case we should know names on all binaries that make produces, setting permissions for them.

so, I guess what you call `/dist` today I was thinking of `/usr/src`; I mean `/dist` as files for distribution.  Basically, install to a synthetic root folder (e.g. `/install`). 

Then we just `COPY --from=builder /install /`

**Comment by alebabai** [comment]
>so, I guess what you call /dist today I was thinking of /usr/src; I mean /dist as files for distribution. Basically, install to a synthetic root folder (e.g. /install).

Not certainly in that way.  For now /dist folder contains src files and all files produced by make.
image.

File: Dockerfile [hunk 3]
```diff
@@ -89,19 +95,13 @@ RUN wget https://github.com/cloudposse/sudosh/releases/download/${SUDOSH_VERSION
...
-COPY --from=google-authenticator-libpam-builder dist dist
-RUN make --directory=dist install && \
-    rm -rf dist
+COPY --from=google-authenticator-libpam-builder dist /
 
 ## Install OpenSSH Portable
-COPY --from=openssh-portable-builder dist dist
-RUN make --directory=dist install && \
-    rm -rf dist
+COPY --from=openssh-portable-builder dist /
```
Comment (osterman) [inline Dockerfile]: use `dist/ /` for symmetry (also a habbit of using `rsync` - you almost always want to match slashes)

File: Dockerfile [hunk 4]
```diff
@@ -89,19 +95,13 @@ RUN wget https://github.com/cloudposse/sudosh/releases/download/${SUDOSH_VERSION
...
-COPY --from=duo-builder dist dist
-RUN make --directory=dist install && \
-    rm -rf dist
+COPY --from=duo-builder dist /
 
 ## Install Google Authenticator PAM module
-COPY --from=google-authenticator-libpam-builder dist dist
-RUN make --directory=dist install && \
-    rm -rf dist
+COPY --from=google-authenticator-libpam-builder dist /
```
Comment (osterman) [inline Dockerfile]: use `dist/ /` for symmetry

File: Dockerfile [hunk 5]
```diff
@@ -89,19 +95,13 @@ RUN wget https://github.com/cloudposse/sudosh/releases/download/${SUDOSH_VERSION
     chmod 755 /usr/bin/sudosh
 
 ## Install Duo
-COPY --from=duo-builder dist dist
-RUN make --directory=dist install && \
-    rm -rf dist
+COPY --from=duo-builder dist /
```
Comment (osterman) [inline Dockerfile]: use `dist/ /` for symmetry

File: Dockerfile [hunk 6]
```diff
@@ -27,12 +30,14 @@ RUN cd dist && \
 FROM builder as google-authenticator-libpam-builder
 
 ARG AUTHENTICATOR_LIBPAM_VERSION=1.05
-RUN git clone --branch ${AUTHENTICATOR_LIBPAM_VERSION} --single-branch https://github.com/google/google-authenticator-libpam dist
+RUN git clone --branch ${AUTHENTICATOR_LIBPAM_VERSION} --single-branch https://github.com/google/google-authenticator-libpam src
 
-RUN cd dist && \
+RUN cd src && \
```
Comment (osterman) [inline Dockerfile]: Question: do we gain anything by chaining commands with `&&` in a multi-stage build where the build step is not shipped in the final image?

Comment (alebabai) [reply Dockerfile]: just looks cleaner and reduces cached layers (disk space)

**Comment by alebabai** [comment]
@osterman it didn't fix issue for me, but make would be useful
So we have `automake` with some standard variables
image

So we can remove redundant params, because prefix is applied to them by default
```
...
        --prefix=/dist/usr \
        --sysconfdir=/etc/ssh \
        --datadir=/usr/share/openssh \
        --libexecdir=/usr/lib/ssh \
        --mandir=/usr/share/man \
...
```
final version could look like this
```
--prefix=/dist/usr \
```

**Comment by alebabai** [comment]
seems to work now

**Comment by osterman** [comment]
```
--prefix=/dist/usr \
```

What about `/etc`? ...don't want `/usr/etc`

**Comment by alebabai** [comment]
> What about /etc? 

We don't need prefix here because this property is used during bastion runtime. So we can use it as is
`--sysconfdir=/etc/ssh`

**Comment by alebabai** [comment]
In current variant i preserved all options, just added prefix where it needed.

**Comment by mhwelsh** [comment]
Pretty sure this Patch breaks duo integration. Current repo version of this does have the duo libs. I built fromscratch and I think it’s because the prefix command doesn’t work for the duo config script.  The libs are built but aren’t copied to the fresh image as intended.

**Comment by osterman** [comment]
@alebabai can you take a look at this on Monday?

**Comment by osterman** [comment]
@mhwelsh can you open an issue so we can track it? thanks!

**Comment by alebabai** [comment]
@osterman
> @alebabai can you take a look at this on Monday?

Sure.

**Pull request merged**